game/src/ungap/route/mod.rs

```diff
@@ -174,40 +207,33 @@ impl State<App> for RoutePlanner {
                 return t;
             }
         }
-        if let Outcome::Changed(ref x) = outcome {
+        if let Outcome::Changed(ref x) = panel_outcome {
             if x == "steep hills" || x == "stressful roads" {
                 app.session.routing_preferences = RoutingPreferences {
                     hills: self.input_panel.dropdown_value("steep hills"),
                     stressful_roads: self.input_panel.dropdown_value("stressful roads"),
                 };
                 self.recalculate_routes(ctx, app);
-                self.update_input_panel(ctx, app);
                 return Transition::Keep;
             }
         }
         // Send all other outcomes here
         // TODO This routing of outcomes and the brittle ordering totally breaks encapsulation :(
         if let Some(t) = self
             .main_route
-            .event(ctx, app, &outcome, &mut self.input_panel)
+            .event(ctx, app, &panel_outcome, &mut self.input_panel)
         {
             return t;
         }
-        // Dragging behavior inside here only works if we're not hovering on an alternate route
-        // TODO But then that prevents dragging some waypoints! Can we give waypoints precedence
-        // instead?
-        if !focused_on_alt_route && self.waypoints.event(ctx, app, outcome) {
+
+        if self
```

Comment:
I'd still like to find a better pattern for delegating panel/world outcomes to the right piece of code. Maybe some kind of "tagged union" thing; a way to up-front mark outcomes as belonging to different pieces of state.